File: 0905-sort-array-by-parity/0905-sort-array-by-parity.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void swap(int* left , int* right){
    int temp = *left;
    *left = *right;
    *right = temp;
}

int* sortArrayByParity(int* nums, int numsSize, int* returnSize) {
    *returnSize = numsSize;
    int* result = (int*)malloc(numsSize* sizeof(int));
    if(result == NULL){
        return NULL;
    }
    memcpy(result , nums , numsSize * sizeof(int));
    int l = 0;
    int r = numsSize - 1;

    while(l < r){
        if(nums[l] % 2 > nums[r] % 2){
            swap(&nums[l] , &nums[r]);
        }

        if(nums[l] % 2 == 0){
            l++;
        }
        if(nums[r] % 2 != 0){
            r--;
        }
    }
    return nums;
}

File: 0905-sort-array-by-parity/0905-sort-array-by-parity.py
class Solution:
    def sortArrayByParity(self, nums: list[int]) -> list[int]:
        # left = 0
        # right = len(nums) - 1

        # while left < right:
        #     if nums[left] % 2 > nums[right] % 2:
        #         nums[left] , nums[right] = nums[right] , nums[left]
        #     if nums[left] % 2 == 0:
        #         left += 1
        #     if nums[right] % 2 != 0:
        #         right -= 1
        # return nums

        even = []
        odd = []
        for i in nums:
            if i % 2 == 0:
                even.append(i)
            else:
                odd.append(i)
        return even + odd